### DISCUSSION TITLE
Are there any downsides or limitations of using Next.js api routes instead of custom server?

### DISCUSSION THREAD

**Discussion by mimoid-prog** [vercel/next.js#17679] [Help] [answered]
I have a huge backend application using Express, with many routes, middlewares, databases etc. Is it possible to recreate all of these features and others using Next.js API routes and is it worth to do so? What are the downsides of using API routes?

**Comment by francisrod01** [comment]
In case you decide to use a custom server, you'll implement your own route patterns.

> A custom server can not be deployed on Vercel, the platform Next.js was made for.

> Before deciding to use a custom server please keep in mind that it should only be used when the integrated router of Next.js can't meet your app requirements. A **custom server** will remove important performance optimizations, like **serverless functions** and [**Automatic Static Optimization**](https://nextjs.org/docs/advanced-features/automatic-static-optimization).

More info: https://nextjs.org/docs/advanced-features/custom-server

**Reply by mimoid-prog** [reply]
@francisrod01 Yeah, I know about the custom server and it's downsides but im curious if I can achieve the same thing by using API routes and If I won't be limited in same way when my backend will get more complex.

**Comment by lfades** [comment, marked as answer]
Serverless functions scale better, are faster, and cost less. The only limitation afaik is that you can't have live connections (like websockets), because that would require the serverless function to stay alive.

**Reply by mimoid-prog** [reply]
Thanks for the answer. But what about db connection? It will be creating new db connection on each request to my api route?

**Reply by lfades** [reply]
Yes. DBs are okay as long as you can either create a huge amount of connections or use a DB that doesn't require a live connection.

**Reply by mimoid-prog** [reply]
I think this is kinda huge problem when you build large application with many api routes. That's why I asked this question and I think it would be nice if there would be a section in the API routes docs with comparison between e.g normal express server and next.js api routes, with pros and cons.

**Reply by andrzejchmura** [reply]
One thing to keep in mind, if you are planning to deploy on Vercel, is that there are limitations on the number of serverless functions allowed per deployment: https://vercel.com/docs/platform/limits

**Reply by timneutkens** [reply]
> One thing to keep in mind, if you are planning to deploy on Vercel, is that there are limitations on the number of serverless functions allowed per deployment: [vercel.com/docs/platform/limits](https://vercel.com/docs/platform/limits)

It's very unlikely you'll run into these when using Next.js as in general only 1 function is deployed for API routes.

**Reply by andrzejchmura** [reply]
That's awesome! Was that always the case? I remember having a hard time deciding if this limitation wont affect me too much, but now I see additional note added to /limits page. Anyway, that's great to hear.

**Reply by nishant-k1** [reply]
Very helpful, Thanks.

**Reply by wizzymore** [reply]
I know this is an old thread but this was the first result when searching for next.js api limitations and i want to address one issue @andrzejchmura mentioned, based on what vercel says on the link provided i don't think you will ever reach the limit on a normal project with the hobby plan:

> Note: Next.js bundles all API Routes and server-rendered pages into shared Serverless Functions , separating API routes and server-rendered pages when deployed on Vercel. Further Serverless Functions are created only if they do not fit within the 50MB limit. As a result, the "Serverless Functions per Deployment" limit is unlikely to apply for Next.js projects.

**Comment by bhatvikrant** [comment]
Note: Really sorry if it's a silly question (hehe) I am trying to learn nextjs ☺️

**Comment by j-lee8** [comment]
I'd be keen to know this. Is it suitable for a social media-like application using GraphQL? Will need posts, comments, upvotes, real time notifications etc. I know Lambdas don't maintain the sockets after cool down, but it seems like such an efficient way to build and deploy. What's the alternative? Spin up a separate AppSync / Apollo server to handle subscriptions?

**Reply by alphaofficial** [reply]
Using api routes as proxy to hasura for something similar. The performance is good but Very unpredictable. I use tanstack react-query with codegen for typed hooks. The Api requests to the graphql randomly fails sometimes.  Still trying to figure out why. Might move to a separate graphql server

**Reply by sgrund14** [reply]
@alphaofficial hey, I see the thing where the API requests fail randomly sometimes too... ever figure it out? my guess was that it was sometimes making too many subrequests, breaking the subrequest limit of the api function https://community.cloudflare.com/t/the-maximum-number-of-concurrent-executions-on-cloudflare-pages/535120/5

thinking about moving to a separate gql server myself (want support for sockets too, anyways)

**Comment by svobik7** [comment]
Hi @mimoid-prog, how did your transition to api routes go? Would you mind to share your experience? Especially with db connections 🙂
